subsys/bluetooth/host/id.c

```diff
@@ -788,13 +792,70 @@ void bt_id_pending_keys_update(void)
 	}
 }
 
+struct bt_id_conflict {
+	struct bt_keys *candidate;
+	struct bt_keys *found;
+};
+
+/* The Controller Resolve List is constrained by 7.8.38 "LE Add Device To
+ * Resolving List command". The Host is designed with the assumption that all
+ * local bonds can be put in the resolve list if there is room. Therefore we
+ * must refuse bonds that conflict in the resolve list. Notably, this prevents
+ * multiple local identities to bond with the same remote identity.
+ */
+void find_rl_conflict(struct bt_keys *resident, void *user_data)
+{
+	struct bt_id_conflict *conflict = user_data;
+	bool addr_conflict;
+	bool irk_conflict;
+
+	__ASSERT_NO_MSG(conflict != NULL);
+	__ASSERT_NO_MSG(conflict->candidate != NULL);
+	__ASSERT_NO_MSG(resident != NULL);
+	/* Only uncommitted bonds can be in conflict with committed bonds. */
+	__ASSERT_NO_MSG((conflict->candidate->state & BT_KEYS_ID_ADDED) == 0);
+
+	if (conflict->found) {
+		return;
+	}
+
+	/* Test against commited bonds only. */
+	if ((resident->state & BT_KEYS_ID_ADDED) == 0) {
+		return;
+	}
+
+	addr_conflict = bt_addr_le_eq(&conflict->candidate->addr, &resident->addr);
+
+	/* All-zero IRK is "no IRK", and does not conflict with other Zero-IRKs. */
+	irk_conflict = (!bt_irk_eq(&conflict->candidate->irk, &(struct bt_irk){})
+			&& bt_irk_eq(&conflict->candidate->irk, &resident->irk));
+
+	if (addr_conflict || irk_conflict) {
+		conflict->found = resident;
+	}
+}
+
+struct bt_keys *bt_id_find_conflict(struct bt_keys *candidate)
+{
+	struct bt_id_conflict conflict = {
+		.candidate = candidate,
+	};
+
+	bt_keys_foreach(BT_KEYS_IRK, find_rl_conflict, &conflict);
+
+	return conflict.found;
+}
+
 void bt_id_add(struct bt_keys *keys)
 {
 	struct bt_conn *conn;
 	int err;
 
 	BT_DBG("addr %s", bt_addr_le_str(&keys->addr));
 
+	__ASSERT_NO_MSG(keys != NULL);
+	__ASSERT_NO_MSG(!bt_id_find_conflict(keys));
```

Comment:
I think i'd rather log an error and exit. Don't wanna have an assert failing in the wild because of something we missed.
We can clean up the bonding procedure later.